File: _archive/obsidian-canvas-learning/_v1-ref/src/services/skill-registry.ts
/**
 * Canvas Learning System - Skill Registry
 * Story 3.5: /Command Skill Integration (AC-1, AC-3, AC-4, Task 2)
 *
 * Scans the `.claude/commands/` directory to discover skill prompt templates.
 * Parses each `.md` file's frontmatter for name, description, and icon.
 * Provides a reactive skill list for the SkillSelector UI to consume.
 *
 * Key behaviors:
 * - Scans `.claude/commands/` in the project root (not Obsidian vault)
 * - Supports user-added custom skills (AC-4: add .md file, no restart needed)
 * - Re-scans on every '/' trigger (lightweight, directory listing only)
 *
 * [Source: _bmad-output/implementation-artifacts/3-5-skill-command-integration.md#Task 2]
 */

import { existsSync, readdirSync, readFileSync } from 'fs';
import { join, basename } from 'path';

// ═══════════════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════════════

export interface Skill {
  /** The command name (file stem without .md). Used as `/command-name`. */
  commandName: string;

  /** Display name from frontmatter or derived from filename. */
  name: string;

  /** Short description from frontmatter. */
  description: string;

  /** Emoji icon from frontmatter (default: tool icon). */
  icon: string;

  /** Absolute path to the .md file. */
  filePath: string;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Default skill metadata (for existing commands without full frontmatter)
// ═══════════════════════════════════════════════════════════════════════════════

const DEFAULT_SKILL_META: Record<string, { name: string; description: string; icon: string }> = {
  'basic-decompose': {
    name: '基础拆解',
    description: '将复杂概念分解为可理解的子概念',
    icon: '🧩',
  },
  'deep-decompose': {
    name: '深度拆解',
    description: '递归分解到原子级知识点，暴露理解盲点',
    icon: '🔬',
  },
  'four-level': {
    name: '四级解释',
    description: '新手/进阶/专家/创新四个层次解释',
    icon: '📊',
  },
  compare: {
    name: '概念对比',
    description: '对比两个容易混淆的概念',
    icon: '⚖️',
  },
  'memory-anchor': {
    name: '记忆锚点',
    description: '为概念创建记忆术/类比/助记',
    icon: '⚓',
  },
  'deep-explain': {
    name: '深度解释',
    description: '深入浅出解释概念的本质原理',
    icon: '💡',
  },
  'example-teach': {
    name: '举例教学',
    description: '通过具体实例理解抽象概念',
    icon: '📝',
  },
  'oral-explain': {
    name: '口语化解释',
    description: '用最通俗的语言解释专业概念',
    icon: '🗣️',
  },
  'question-decompose': {
    name: '问题拆解',
    description: '将大问题分解为可回答的小问题',
    icon: '❓',
  },
  score: {
    name: '知识检验',
    description: '快速检测对概念的理解程度',
    icon: '✅',
  },
  'verify-question': {
    name: '验证追问',
    description: '通过追问验证理解的深度',
    icon: '🔍',
  },
};

// ═══════════════════════════════════════════════════════════════════════════════
// SkillRegistry
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Manages the registry of available skills.
 *
 * Skills are discovered from `.claude/commands/` directory.
 * The list is refreshed on each '/' trigger to pick up newly added skills
 * without requiring a plugin restart (AC-4).
 */
class SkillRegistry {
  /** Skill list. */
  skills: Skill[] = [];

  /** Cached commands directory path. */
  private commandsDir: string | null = null;

  /**
   * Scan and load all skills from the commands directory.
   *
   * @param projectRoot - The project root directory containing `.claude/commands/`.
   *   For this project, the vault base path is used to navigate to the project root.
   */
  loadSkills(projectRoot: string): void {
    // The commands directory is at the project root level
    // Navigate from vault path to find the .claude/commands directory
    const candidates = [
      join(projectRoot, '.claude', 'commands'),
      // Also check parent directories in case vault is nested
      join(projectRoot, '..', 'canvas-learning-system', '.claude', 'commands'),
      join(projectRoot, '..', '..', 'canvas-learning-system', '.claude', 'commands'),
    ];

    let commandsDir: string | null = null;
    for (const candidate of candidates) {
      if (existsSync(candidate)) {
        commandsDir = candidate;
        break;
      }
    }

    if (!commandsDir) {
      console.warn(
        '[Canvas Learning] SkillRegistry: .claude/commands/ directory not found',
      );
      this.skills = [];
      return;
    }

    this.commandsDir = commandsDir;
    this.scanDirectory();
  }

  /**
   * Rescan the commands directory for skill files.
   * Called on every '/' trigger to pick up new files (AC-4).
   */
  refresh(): void {
    if (this.commandsDir) {
      this.scanDirectory();
    }
  }

  /**
   * Scan the directory and parse each .md file.
   */
  private scanDirectory(): void {
    if (!this.commandsDir || !existsSync(this.commandsDir)) {
      this.skills = [];
      return;
    }

    const files = readdirSync(this.commandsDir).filter((f) =>
      f.endsWith('.md'),
    );

    const skills: Skill[] = [];

    for (const file of files) {
      const filePath = join(this.commandsDir, file);
      const commandName = basename(file, '.md');
      const skill = this.parseSkillFile(filePath, commandName);
      if (skill) {
        skills.push(skill);
      }
    }

    // Sort alphabetically by name
    skills.sort((a, b) => a.name.localeCompare(b.name, 'zh-CN'));
    this.skills = skills;
  }

  /**
   * Parse a single skill .md file.
   *
   * Extracts frontmatter fields: name, description, icon.
   * Falls back to DEFAULT_SKILL_META or derived values.
   */
  private parseSkillFile(filePath: string, commandName: string): Skill | null {
    try {
      const content = readFileSync(filePath, 'utf-8');
      const frontmatter = this.parseFrontmatter(content);
      const defaults = DEFAULT_SKILL_META[commandName];

      return {
        commandName,
        name:
          frontmatter.name ??
          defaults?.name ??
          commandName.replace(/-/g, ' '),
        description:
          frontmatter.description ??
          defaults?.description ??
          '',
        icon:
          frontmatter.icon ??
          defaults?.icon ??
          '🔧',
        filePath,
      };
    } catch (err) {
      console.warn(
        `[Canvas Learning] SkillRegistry: failed to parse ${filePath}:`,
        err,
      );
      return null;
    }
  }

  /**
   * Parse YAML-like frontmatter from a Markdown file.
   * Simple parser for key: value pairs between --- delimiters.
   */
  private parseFrontmatter(content: string): Record<string, string> {
    const result: Record<string, string> = {};
    const match = content.match(/^---\s*\n([\s\S]*?)\n---/);
    if (!match) return result;

    const lines = match[1].split('\n');
    for (const line of lines) {
      const colonIdx = line.indexOf(':');
      if (colonIdx < 0) continue;

      const key = line.slice(0, colonIdx).trim();
      let value = line.slice(colonIdx + 1).trim();

      // Skip array values (like allowed-tools)
      if (value.startsWith('[') || value === '') continue;

      // Remove quotes
      if (
        (value.startsWith('"') && value.endsWith('"')) ||
        (value.startsWith("'") && value.endsWith("'"))
      ) {
        value = value.slice(1, -1);
      }

      result[key] = value;
    }

    return result;
  }
}

/** Singleton skill registry instance. */
export const skillRegistry = new SkillRegistry();

File: backend/app/services/skill_registry.py
# Story 3.5: /命令技能集成 — SkillRegistry Service
# [Source: _bmad-output/implementation-artifacts/3-5-skill-command-integration.md]
"""
SkillRegistry — Service for scanning .claude/commands/ directory
and exposing skill metadata to the frontend via API.

Responsibilities:
  - Scan .claude/commands/ directory for all .md skill template files
  - Parse YAML frontmatter (name, description, icon) if present
  - Fallback: extract name from first # heading, description from first paragraph
  - Provide list_skills() API returning structured skill metadata
  - Provide get_skill_content() API returning full template content for execution
  - Support dynamic refresh (rescan on demand, no restart required)

Usage:
    registry = SkillRegistry.get_instance()
    skills = registry.list_skills()           # All skills with metadata
    content = registry.get_skill_content("basic-decompose")  # Full template
"""

import asyncio
import logging
import re
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional

logger = logging.getLogger(__name__)

# Default commands directory — project root / .claude / commands
_CONFIG_DIR = Path(__file__).parent.parent  # backend/app/
_BACKEND_DIR = _CONFIG_DIR.parent  # backend/
_PROJECT_ROOT = _BACKEND_DIR.parent  # project root
_DEFAULT_COMMANDS_DIR = _PROJECT_ROOT / ".claude" / "commands"

# Regex for YAML frontmatter extraction
_FRONTMATTER_PATTERN = re.compile(
    r"^---\s*\n(.*?)\n---\s*\n",
    re.DOTALL,
)

# Regex for extracting key-value pairs from YAML (simple flat YAML only)
_YAML_KV_PATTERN = re.compile(r"^(\w[\w-]*)\s*:\s*(.+)$", re.MULTILINE)

# Regex for first Markdown heading
_HEADING_PATTERN = re.compile(r"^#\s+(.+)$", re.MULTILINE)


@dataclass(frozen=True)
class SkillInfo:
    """Immutable representation of a skill command template."""

    skill_id: str  # File stem, e.g., "basic-decompose"
    name: str  # Display name, e.g., "基础拆解"
    description: str  # One-line description
    icon: str  # Icon identifier (e.g., "puzzle", "target")
    file_path: str  # Absolute path to the .md file


@dataclass
class SkillContent:
    """Skill metadata plus full template content."""

    skill_id: str
    name: str
    description: str
    icon: str
    content: str  # Full .md file content (including frontmatter)
    file_path: str


class SkillRegistry:
    """
    Registry that scans .claude/commands/ and provides skill metadata.

    Thread-safe. Call load() once during startup or on-demand via refresh().
    """

    _instance: Optional["SkillRegistry"] = None
    # Use threading.Lock for the synchronous singleton pattern (called from
    # both sync and async code paths). The lock is only held for object
    # construction (microseconds), so it does not block the event loop.
    _lock = threading.Lock()

    def __init__(self, commands_dir: Optional[Path] = None):
        self._commands_dir = commands_dir or _DEFAULT_COMMANDS_DIR
        self._skills: Dict[str, SkillInfo] = {}
        self._loaded = False
        # asyncio.Lock for protecting concurrent async operations (e.g. refresh)
        self._async_lock = asyncio.Lock()

    # ─── Singleton ────────────────────────────────────────────────────────

    @classmethod
    def get_instance(cls, commands_dir: Optional[Path] = None) -> "SkillRegistry":
        """Return the singleton SkillRegistry, creating it if needed.

        Uses threading.Lock (not asyncio.Lock) because this is a synchronous
        class method that must work from both sync and async contexts.
        The lock is held only for object construction (microseconds).
        """
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = cls(commands_dir=commands_dir)
        return cls._instance

    @classmethod
    def reset_instance(cls) -> None:
        """Reset the singleton (for testing only)."""
        with cls._lock:
            cls._instance = None

    # ─── Loading ──────────────────────────────────────────────────────────

    def load(self) -> int:
        """
        Scan .claude/commands/ and load all .md skill templates.

        Returns:
            Number of skills successfully loaded.
        """
        if not self._commands_dir.exists():
            logger.warning(
                "[SkillRegistry] Commands directory does not exist: %s",
                self._commands_dir,
            )
            self._skills.clear()
            self._loaded = True
            return 0

        self._skills.clear()
        loaded_count = 0

        for md_file in sorted(self._commands_dir.glob("*.md")):
            try:
                skill = self._parse_skill(md_file)
                self._skills[skill.skill_id] = skill
                loaded_count += 1
                logger.debug(
                    "[SkillRegistry] Loaded skill: %s (%s)",
                    skill.skill_id,
                    skill.name,
                )
            except (OSError, ValueError, UnicodeDecodeError) as exc:
                logger.warning(
                    "[SkillRegistry] Failed to parse skill file %s: %s",
                    md_file.name,
                    exc,
                )

        self._loaded = True
        logger.info(
            "[SkillRegistry] Loaded %d skills from %s",
            loaded_count,
            self._commands_dir,
        )
        return loaded_count

    def refresh(self) -> int:
        """
        Rescan the commands directory. Supports AC-4 (user custom skills).

        Returns:
            Number of skills after refresh.
        """
        return self.load()

    def _parse_skill(self, file_path: Path) -> SkillInfo:
        """Parse a single .md file into SkillInfo."""
        raw = file_path.read_text(encoding="utf-8")
        skill_id = file_path.stem  # e.g., "basic-decompose"

        name = ""
        description = ""
        icon = "file-text"  # default icon

        # Try extracting from YAML frontmatter.
        # Use yaml.safe_load for robust parsing of multi-line values, quoted
        # strings, and other YAML features that the simple regex misses.
        fm_match = _FRONTMATTER_PATTERN.match(raw)
        if fm_match:
            fm_block = fm_match.group(1)
            try:
                import yaml

                fm_data = yaml.safe_load(fm_block) or {}
                if isinstance(fm_data, dict):
                    name = str(fm_data.get("name", "")).strip()
                    description = str(fm_data.get("description", "")).strip()
                    icon = str(fm_data.get("icon", icon)).strip()
                else:
                    # YAML parsed but not a dict — fall through to heading
                    fm_data = {}
            except Exception:
                # yaml.safe_load failed — fallback to regex key-value extraction
                fm_kvs = dict(_YAML_KV_PATTERN.findall(fm_block))
                name = fm_kvs.get("name", "").strip().strip('"').strip("'")
                description = fm_kvs.get("description", "").strip().strip('"').strip("'")
                icon = fm_kvs.get("icon", icon).strip().strip('"').strip("'")

        # Fallback: extract name from first # heading
        if not name:
            heading_match = _HEADING_PATTERN.search(raw)
            if heading_match:
                name = heading_match.group(1).strip()

        # Fallback: use skill_id as name
        if not name:
            name = skill_id

        # Fallback: extract description from first non-heading paragraph
        if not description:
            description = self._extract_first_paragraph(raw)

        return SkillInfo(
            skill_id=skill_id,
            name=name,
            description=description,
            icon=icon,
            file_path=str(file_path.resolve()),
        )

    @staticmethod
    def _extract_first_paragraph(content: str) -> str:
        """
        Extract the first non-empty paragraph after frontmatter and heading.

        Used as fallback description when frontmatter lacks 'description'.
        """
        # Remove frontmatter
        fm_match = _FRONTMATTER_PATTERN.match(content)
        if fm_match:
            content = content[fm_match.end() :]

        lines = content.split("\n")
        paragraph_lines: list[str] = []
        found_heading = False

        for line in lines:
            stripped = line.strip()
            # Skip empty lines before content
            if not stripped:
                if paragraph_lines:
                    break  # End of first paragraph
                continue
            # Skip headings
            if stripped.startswith("#"):
                found_heading = True
                continue
            # Skip usage lines and code blocks
            if stripped.startswith("**用法**") or stripped.startswith("```"):
                if paragraph_lines:
                    break
                continue
            # Collect paragraph text
            if found_heading or not stripped.startswith("#"):
                paragraph_lines.append(stripped)

        result = " ".join(paragraph_lines)
        # Truncate to reasonable description length
        if len(result) > 120:
            result = result[:117] + "..."
        return result

    # ─── Query API ────────────────────────────────────────────────────────

    def list_skills(self) -> List[SkillInfo]:
        """
        Return all loaded skills as a sorted list.

        Auto-loads if not yet loaded.
        """
        if not self._loaded:
            self.load()
        return sorted(self._skills.values(), key=lambda s: s.skill_id)

    def get_skill(self, skill_id: str) -> Optional[SkillInfo]:
        """Return skill metadata by ID, or None if not found."""
        if not self._loaded:
            self.load()
        return self._skills.get(skill_id)

    def get_skill_content(self, skill_id: str) -> Optional[SkillContent]:
        """
        Return full skill metadata + template content.

        Reads the file content fresh each time to support live editing.
        """
        skill = self.get_skill(skill_id)
        if skill is None:
            return None

        file_path = Path(skill.file_path)
        if not file_path.exists():
            logger.warning(
                "[SkillRegistry] Skill file no longer exists: %s",
                skill.file_path,
            )
            return None

        content = file_path.read_text(encoding="utf-8")
        return SkillContent(
            skill_id=skill.skill_id,
            name=skill.name,
            description=skill.description,
            icon=skill.icon,
            content=content,
            file_path=skill.file_path,
        )

    @property
    def is_loaded(self) -> bool:
        """Whether load() has been called."""
        return self._loaded

    @property
    def commands_dir(self) -> Path:
        """Return the configured commands directory path."""
        return self._commands_dir


# ─── Module-level convenience ─────────────────────────────────────────────


def get_skill_registry(
    commands_dir: Optional[Path] = None,
) -> SkillRegistry:
    """Get the singleton SkillRegistry instance."""
    return SkillRegistry.get_instance(commands_dir=commands_dir)
